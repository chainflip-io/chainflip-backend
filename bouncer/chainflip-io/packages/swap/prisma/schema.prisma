// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

// `pnpm prisma format` after changes :)

generator client {
  provider        = "prisma-client-js"
  // https://github.com/prisma/prisma/issues/2584#issuecomment-646414549
  // https://github.com/prisma/prisma/issues/6603#issuecomment-849979849
  output          = "../node_modules/.prisma/client"
  previewFeatures = ["multiSchema"]
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
  schemas  = ["public", "private"]
}

model DepositChannel {
  id             BigInt  @id @default(autoincrement())
  channelId      BigInt
  srcChain       Chain
  issuedBlock    Int
  depositAddress String
  isSwapping     Boolean

  @@unique([issuedBlock, srcChain, channelId])
  @@index([depositAddress])
  @@schema("private")
}

enum Chain {
  Polkadot
  Ethereum
  Bitcoin

  @@schema("public")
}

enum Asset {
  FLIP
  USDC
  DOT
  ETH
  BTC

  @@schema("public")
}

model SwapDepositChannel {
  id                    BigInt       @id @default(autoincrement())
  channelId             BigInt
  srcChain              Chain
  srcAsset              Asset
  depositAddress        String
  expectedDepositAmount Decimal?     @db.Decimal(30, 0)
  destAsset             Asset
  destAddress           String
  ccmGasBudget          Decimal?     @db.Decimal(30, 0)
  ccmMessage            String?
  issuedBlock           Int
  expiryBlock           Int? // TODO(mainnet): remove this column
  srcChainExpiryBlock   BigInt?
  brokerCommissionBps   Int
  isExpired             Boolean      @default(false)
  createdAt             DateTime     @default(now())
  swaps                 Swap[]
  failedSwaps           FailedSwap[]
  estimatedExpiryAt     DateTime?
  openedThroughBackend  Boolean      @default(false)

  @@unique([issuedBlock, srcChain, channelId])
  @@index([depositAddress])
  @@schema("public")
}

enum SwapType {
  SWAP
  PRINCIPAL
  GAS

  @@schema("public")
}

model Swap {
  id                           BigInt              @id @default(autoincrement())
  /// the swap id on the state chain
  nativeId                     BigInt              @unique
  type                         SwapType
  srcAsset                     Asset
  depositAmount                Decimal             @db.Decimal(30, 0)
  swapInputAmount              Decimal             @db.Decimal(30, 0) // deposit amount after ingress fee was decucted
  intermediateAmount           Decimal?            @db.Decimal(30, 0)
  swapOutputAmount             Decimal?            @db.Decimal(30, 0) // output of the swap before egress fee is deducted
  fees                         SwapFee[]
  destAsset                    Asset
  destAddress                  String
  depositReceivedAt            DateTime
  depositReceivedBlockIndex    String
  swapExecutedAt               DateTime?
  swapExecutedBlockIndex       String?
  swapDepositChannel           SwapDepositChannel? @relation(fields: [swapDepositChannelId], references: [id])
  swapDepositChannelId         BigInt?
  ccmDepositReceivedBlockIndex String?
  ccmGasBudget                 Decimal?            @db.Decimal(30, 0)
  ccmMessage                   String?
  txHash                       String?
  egress                       Egress?             @relation(fields: [egressId], references: [id])
  egressId                     BigInt?
  createdAt                    DateTime            @default(now())
  updatedAt                    DateTime            @default(now()) @updatedAt

  @@schema("public")
}

enum SwapFeeType {
  LIQUIDITY
  NETWORK
  INGRESS
  EGRESS
  BROKER

  @@schema("public")
}

model SwapFee {
  id     BigInt      @id @default(autoincrement())
  swap   Swap        @relation(fields: [swapId], references: [id])
  swapId BigInt
  type   SwapFeeType
  asset  Asset
  amount Decimal     @db.Decimal(30, 0)

  @@index([swapId])
  @@schema("public")
}

enum FailedSwapType {
  FAILED
  IGNORED

  @@schema("public")
}
enum FailedSwapReason {
  BelowMinimumDeposit 
  NotEnoughToPayFees // gas price to process deposit is too high
  EgressAmountZero  // egress amount left after fees is too low 
  
  @@schema("public")
}
model FailedSwap {
  id                   Int                 @id @default(autoincrement())
  type                 FailedSwapType
  reason               FailedSwapReason?
  destAddress          String
  destChain            Chain
  depositAmount        Decimal             @db.Decimal(30, 0)
  srcChain             Chain
  swapDepositChannel   SwapDepositChannel? @relation(fields: [swapDepositChannelId], references: [id])
  swapDepositChannelId BigInt?
  txHash               String?

  @@index([swapDepositChannelId])
  @@schema("public")
}

model Egress {
  id                  BigInt     @id @default(autoincrement())
  /// the egress id on the state chain
  nativeId            BigInt
  chain               Chain
  amount              Decimal    @db.Decimal(30, 0)
  scheduledAt         DateTime
  scheduledBlockIndex String
  swaps               Swap[]
  broadcast           Broadcast? @relation(fields: [broadcastId], references: [id])
  broadcastId         BigInt?
  createdAt           DateTime   @default(now())
  updatedAt           DateTime   @default(now()) @updatedAt

  @@unique([nativeId, chain])
  @@schema("public")
}

enum BroadcastType {
  BATCH
  CCM

  @@schema("public")
}

model Broadcast {
  id                  BigInt        @id @default(autoincrement())
  /// the broadcast id on the state chain
  nativeId            BigInt
  chain               Chain
  type                BroadcastType @default(BATCH)
  egresses            Egress[]
  requestedAt         DateTime
  requestedBlockIndex String
  succeededAt         DateTime?
  succeededBlockIndex String?
  abortedAt           DateTime?
  abortedBlockIndex   String?
  replacedBy          Broadcast?    @relation("replacedBy", fields: [replacedById], references: [id])
  replaces            Broadcast?    @relation("replacedBy")
  replacedById        BigInt?       @unique
  createdAt           DateTime      @default(now())
  updatedAt           DateTime      @default(now()) @updatedAt

  @@unique([nativeId, chain])
  @@schema("public")
}

model State {
  id     Int @id @default(autoincrement())
  height Int @default(0)

  @@schema("private")
}

model MarketMaker {
  id        Int    @id @default(autoincrement())
  name      String @unique
  publicKey String

  @@schema("private")
}

enum ThirdPartyProtocol {
  lifi
  squid

  @@schema("public")
}

model ThirdPartySwap {
  id            BigInt             @id @default(autoincrement())
  uuid          String             @unique
  protocol      ThirdPartyProtocol
  routeResponse Json
  txHash        String
  txLink        String
  createdAt     DateTime           @default(now())

  @@schema("public")
}

model Pool {
  id                        Int   @id @default(autoincrement())
  baseAsset                 Asset
  quoteAsset                Asset
  liquidityFeeHundredthPips Int

  @@unique([baseAsset, quoteAsset])
  @@schema("public")
}

model ChainTracking {
  id             Int      @id @default(autoincrement())
  chain          Chain    @unique
  height         BigInt   @default(0)
  blockTrackedAt DateTime @default(now())
  updatedAt      DateTime @updatedAt @db.Timestamptz(3)

  @@schema("public")
}
