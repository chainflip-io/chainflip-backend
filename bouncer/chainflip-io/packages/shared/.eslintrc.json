{
  "extends": "../../.eslintrc.json",
  "plugins": ["eslint-plugin-n"],
  "rules": {
    "n/no-process-env": ["error"]
  },
  "overrides": [
    {
      "files": ["*.test.ts", "*.mjs", "*.js", "*.cjs"],
      "rules": {
        "n/no-process-env": ["off"]
      }
    }
  ]
}
