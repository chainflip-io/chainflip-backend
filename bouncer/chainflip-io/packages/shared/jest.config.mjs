// eslint-disable-next-line import/no-relative-packages
import baseConfig from '../../jest.config.mjs';

process.env.RPC_BROKER_HTTPS_URL = 'wss://rpc-broker.example.com';

export default {
  ...baseConfig,
};
