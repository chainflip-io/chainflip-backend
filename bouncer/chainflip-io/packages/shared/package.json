{
  "name": "@chainflip-io/sdk-shared",
  "version": "1.0.0",
  "description": "",
  "scripts": {
    "typecheck": "tsc --noEmit",
    "clean": "rm -rf dist",
    "build": "tsc -p tsconfig.cjs.json && tsc -p tsconfig.mjs.json",
    "build:abis": "typechain --target=ethers-v6 --out-dir=./src/abis ./abis/**",
    "lint": "pnpm eslint --max-warnings 0 './**/*.ts'",
    "test": "jest",
    "test:ci": "CI=1 pnpm test -- --forceExit"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "dependencies": {
    "@polkadot/util": "^12.6.2",
    "@polkadot/util-crypto": "^12.6.2",
    "axios": "^1.6.7",
    "bech32-buffer": "^0.2.1",
    "ethers": "^6.9.2",
    "ioredis": "^5.3.2",
    "rxjs": "^7.8.1",
    "ws": "^8.16.0",
    "zod": "^3.22.4"
  },
  "devDependencies": {
    "@ethersproject/abi": "^5.7.0",
    "@ethersproject/providers": "^5.7.2",
    "@typechain/ethers-v6": "^0.5.1",
    "@types/ws": "^8.5.10",
    "eslint-plugin-n": "^16.6.2",
    "typechain": "^8.3.2",
    "winston": "^3.11.0"
  }
}
