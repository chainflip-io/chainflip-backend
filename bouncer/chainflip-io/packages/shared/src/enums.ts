type ArrayToMap<T extends readonly string[]> = {
  [K in T[number]]: K;
};

const arrayToMap = <const T extends readonly string[]>(
  array: T,
): ArrayToMap<T> =>
  Object.fromEntries(array.map((key) => [key, key])) as ArrayToMap<T>;

export const Chains = arrayToMap(['Bitcoin', 'Ethereum', 'Polkadot', 'Arbitrum']);
export type Chain = (typeof Chains)[keyof typeof Chains];

export const Assets = arrayToMap(['FLIP', 'USDC', 'DOT', 'ETH', 'BTC', 'ARBETH', 'ARBUSDC']);
export type Asset = (typeof Assets)[keyof typeof Assets];

export const ChainflipNetworks = arrayToMap([
  'sisyphos',
  'perseverance',
  'mainnet',
  'partnernet',
]);

export type ChainflipNetwork =
  (typeof ChainflipNetworks)[keyof typeof ChainflipNetworks];

export const isTestnet = (network: ChainflipNetwork): boolean =>
  network !== ChainflipNetworks.mainnet;

export const assetChains = {
  [Assets.ETH]: Chains.Ethereum,
  [Assets.FLIP]: Chains.Ethereum,
  [Assets.USDC]: Chains.Ethereum,
  [Assets.BTC]: Chains.Bitcoin,
  [Assets.DOT]: Chains.Polkadot,
  [Assets.ARBETH]: Chains.Arbitrum,
  [Assets.ARBUSDC]: Chains.Arbitrum,
} satisfies Record<Asset, Chain>;

export const assetDecimals = {
  [Assets.DOT]: 10,
  [Assets.ETH]: 18,
  [Assets.FLIP]: 18,
  [Assets.USDC]: 6,
  [Assets.BTC]: 8,
  [Assets.ARBETH]: 18,
  [Assets.ARBUSDC]: 6,
} satisfies Record<Asset, number>;

// https://github.com/chainflip-io/chainflip-backend/blob/a2a3c2e447e7b629c4b96797d9eed22eb5b87a0b/state-chain/primitives/src/chains/assets.rs#L51-L59
export const assetContractIds: Record<Asset, number> = {
  // 0 is reservered for particular cross chain messaging scenarios where we want to pass
  // through a message without making a swap.
  [Assets.ETH]: 1,
  [Assets.FLIP]: 2,
  [Assets.USDC]: 3,
  [Assets.DOT]: 4,
  [Assets.BTC]: 5,
  [Assets.ARBETH]: 6,
  [Assets.ARBUSDC]: 7,
};

export const chainAssets = {
  [Chains.Ethereum]: [Assets.ETH, Assets.USDC, Assets.FLIP],
  [Chains.Bitcoin]: [Assets.BTC],
  [Chains.Polkadot]: [Assets.DOT],
  [Chains.Arbitrum]: [Assets.ARBETH, Assets.ARBUSDC],
} satisfies Record<Chain, Asset[]>;

// https://github.com/chainflip-io/chainflip-backend/blob/a2a3c2e447e7b629c4b96797d9eed22eb5b87a0b/state-chain/primitives/src/chains.rs#L52-L56
export const chainContractIds: Record<Chain, number> = {
  [Chains.Ethereum]: 1,
  [Chains.Polkadot]: 2,
  [Chains.Bitcoin]: 3,
  [Chains.Arbitrum]: 4,
};
