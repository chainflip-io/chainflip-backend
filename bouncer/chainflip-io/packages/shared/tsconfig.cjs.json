{
  "extends": "./tsconfig",
  "compilerOptions": {
    "module": "CommonJS",
    "outDir": "./dist/cjs" /* Redirect output structure to the directory. */,
    "noEmit": false
  }
}
