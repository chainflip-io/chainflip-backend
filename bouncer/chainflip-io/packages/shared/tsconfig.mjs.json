{
  "extends": "./tsconfig",
  "compilerOptions": {
    "module": "ESNext",
    "outDir": "./dist/mjs" /* Redirect output structure to the directory. */,
    "noEmit": false
  }
}
