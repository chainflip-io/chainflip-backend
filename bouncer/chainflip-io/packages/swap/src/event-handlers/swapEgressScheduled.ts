import { z } from 'zod';
import { Asset, assetChains } from '@/shared/enums';
import { bigintMin } from '@/shared/functions';
import { chainflipChain, u128, unsignedInteger } from '@/shared/parsers';
import { Environment, getEnvironment } from '@/shared/rpc';
import { readAssetValue } from '@/shared/rpc/utils';
import { CacheMap } from '@/swap/utils/dataStructures';
import { estimateIngressEgressFeeAssetAmount } from '@/swap/utils/fees';
import env from '../config/env';
import type { EventHandlerArgs } from '.';

const eventArgsWithoutFee = z.object({
  swapId: unsignedInteger,
  egressId: z.tuple([
    z.object({ __kind: chainflipChain }).transform(({ __kind }) => __kind),
    unsignedInteger,
  ]),
});

const v120EventArgs = eventArgsWithoutFee.and(
  z.object({ amount: u128, fee: u128 }),
);

const eventArgs = z.union([v120EventArgs, eventArgsWithoutFee]);

const environmentByBlockHashCache = new CacheMap<
  string,
  Promise<Environment | null>
>(60_000);

const methodNotFoundRegExp = /Exported method .+ is not found/;
const rpcConfig = { rpcUrl: env.RPC_NODE_HTTP_URL };

const getCachedEnvironmentAtBlock = async (
  blockHash: string,
): Promise<Environment | null> => {
  const cached = environmentByBlockHashCache.get(blockHash);
  if (cached) return cached;

  const environment = getEnvironment(rpcConfig, blockHash).catch((e: Error) => {
    const cause = e.cause as { message: string } | undefined;

    if (methodNotFoundRegExp.test(cause?.message ?? '')) return null;

    environmentByBlockHashCache.delete(blockHash);
    throw e;
  });

  environmentByBlockHashCache.set(blockHash, environment);

  return environment;
};

const assetAmountByNativeAmountAndBlockHashCache = new CacheMap<
  string,
  Promise<bigint>
>(60_000);

const getCachedAssetAmountAtBlock = async (
  asset: Asset,
  nativeAmount: bigint,
  blockHash: string,
): Promise<bigint> => {
  const cacheKey = `${asset}-${nativeAmount.toString()}-${blockHash}`;
  const cached = assetAmountByNativeAmountAndBlockHashCache.get(cacheKey);
  if (cached) return cached;

  const rate = estimateIngressEgressFeeAssetAmount(
    nativeAmount,
    asset,
    blockHash,
  ).catch((e: Error) => {
    assetAmountByNativeAmountAndBlockHashCache.delete(blockHash);
    throw e;
  });

  assetAmountByNativeAmountAndBlockHashCache.set(cacheKey, rate);

  return rate;
};

const getEgressFeeAtBlock = async (
  blockHash: string,
  asset: Asset,
): Promise<bigint> => {
  const environment = await getCachedEnvironmentAtBlock(blockHash);
  if (!environment) return 0n;

  const nativeFee = readAssetValue(environment.ingressEgress.egressFees, {
    asset,
    chain: assetChains[asset],
  });

  return getCachedAssetAmountAtBlock(asset, nativeFee, blockHash);
};

/**
 * this event is emitted in order to correlate the egress id from a network
 * deposit/egress pallet to a swap id
 */
export default async function swapEgressScheduled({
  prisma,
  event,
  block,
}: EventHandlerArgs): Promise<void> {
  const {
    swapId,
    egressId: [chain, nativeId],
    ...restArgs
  } = eventArgs.parse(event.args);

  const swap = await prisma.swap.findUniqueOrThrow({
    where: { nativeId: swapId },
  });

  let egressFee;
  let egress;

  if ('fee' in restArgs) {
    egressFee = restArgs.fee;
    egress = await prisma.egress.create({
      data: {
        nativeId,
        chain,
        amount: restArgs.amount.toString(),
        scheduledAt: new Date(block.timestamp),
        scheduledBlockIndex: `${block.height}-${event.indexInBlock}`,
      },
    });
  } else {
    egressFee = bigintMin(
      await getEgressFeeAtBlock(block.hash, swap.destAsset),
      BigInt(swap.swapOutputAmount?.toFixed() ?? 0),
    );
    egress = await prisma.egress.update({
      where: { nativeId_chain: { chain, nativeId } },
      data: {
        amount: swap.swapOutputAmount?.sub(egressFee.toString()),
      },
    });
  }

  await prisma.swap.update({
    where: { nativeId: swapId },
    data: {
      egress: { connect: { id: egress.id } },
      fees: {
        create: {
          type: 'EGRESS',
          asset: swap.destAsset,
          amount: egressFee.toString(),
        },
      },
    },
  });
}
