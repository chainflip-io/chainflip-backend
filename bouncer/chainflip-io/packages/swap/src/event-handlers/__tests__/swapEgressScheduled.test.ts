import axios from 'axios';
import { Assets } from '@/shared/enums';
import { environment, swapRate } from '@/shared/tests/fixtures';
import {
  DOT_ADDRESS,
  createDepositChannel,
  swapEgressScheduledMock,
  ETH_ADDRESS,
} from './utils';
import prisma from '../../client';
import swapEgressScheduled from '../swapEgressScheduled';

const {
  eventContext: { event },
  block,
} = swapEgressScheduledMock;

jest.mock('axios', () => ({
  post: jest.fn((url, data) => {
    if (data.method === 'cf_environment') {
      return Promise.resolve({ data: environment({ egressFee: '0x777777' }) });
    }

    if (data.method === 'cf_swap_rate') {
      return Promise.resolve({
        data: swapRate({
          output: `0x${(BigInt(data.params[2]) * 3n).toString(16)}`,
        }),
      });
    }

    throw new Error(`unexpected axios call to ${url}: ${JSON.stringify(data)}`);
  }),
}));

describe(swapEgressScheduled, () => {
  beforeEach(async () => {
    await prisma.$queryRaw`TRUNCATE TABLE "SwapDepositChannel", "Swap", "Egress" CASCADE`;
  });

  it('updates an existing swap when the egress asset is the native asset', async () => {
    const { swapId } = event.args;

    await prisma.egress.create({
      data: {
        chain: event.args.egressId[0].__kind,
        nativeId: BigInt(event.args.egressId[1]),
        amount: '1234567890',
        scheduledAt: new Date(block.timestamp),
        scheduledBlockIndex: `${block.height}-${event.indexInBlock}`,
      },
    });

    // store a new swap intent to initiate a new swap
    const swapDepositChannel = await createDepositChannel({
      swaps: {
        create: {
          nativeId: BigInt(swapId),
          depositAmount: '10000000000',
          swapInputAmount: '10000000000',
          swapOutputAmount: '10000000000',
          depositReceivedAt: new Date(block.timestamp - 12000),
          depositReceivedBlockIndex: `${block.height - 100}-${
            event.indexInBlock
          }`,
          swapExecutedAt: new Date(block.timestamp - 6000),
          swapExecutedBlockIndex: `${block.height}-${event.indexInBlock}`,
          srcAsset: Assets.ETH,
          destAsset: Assets.DOT,
          destAddress: DOT_ADDRESS,
          type: 'SWAP',
        },
      },
    });

    await prisma.$transaction((tx) =>
      swapEgressScheduled({
        block: block as any,
        event: event as any,
        prisma: tx,
      }),
    );

    const swap = await prisma.swap.findFirstOrThrow({
      where: { swapDepositChannelId: swapDepositChannel.id },
      include: {
        egress: {
          select: {
            amount: true,
            scheduledAt: true,
            chain: true,
          },
        },
        fees: true,
      },
    });

    expect(swap).toMatchSnapshot({
      id: expect.any(BigInt),
      egressId: expect.any(BigInt),
      createdAt: expect.any(Date),
      updatedAt: expect.any(Date),
      swapDepositChannelId: expect.any(BigInt),
      fees: [{ id: expect.any(BigInt), swapId: expect.any(BigInt) }],
    });
    expect(axios.post).not.toHaveBeenCalledWith(
      expect.anything(),
      expect.objectContaining({ method: 'cf_swap_rate' }),
    );
  });

  it('updates an existing swap when the egress amount is higher than the swapOutputAmount', async () => {
    const { swapId } = event.args;

    await prisma.egress.create({
      data: {
        chain: event.args.egressId[0].__kind,
        nativeId: BigInt(event.args.egressId[1]),
        amount: '1234567890',
        scheduledAt: new Date(block.timestamp),
        scheduledBlockIndex: `${block.height}-${event.indexInBlock}`,
      },
    });

    // store a new swap intent to initiate a new swap
    const swapDepositChannel = await createDepositChannel({
      swaps: {
        create: {
          nativeId: BigInt(swapId),
          depositAmount: '1000',
          swapInputAmount: '1000',
          swapOutputAmount: '1000',
          depositReceivedAt: new Date(block.timestamp - 12000),
          depositReceivedBlockIndex: `${block.height - 100}-${
            event.indexInBlock
          }`,
          swapExecutedAt: new Date(block.timestamp - 6000),
          swapExecutedBlockIndex: `${block.height}-${event.indexInBlock}`,
          srcAsset: Assets.ETH,
          destAsset: Assets.DOT,
          destAddress: DOT_ADDRESS,
          type: 'SWAP',
        },
      },
    });

    await prisma.$transaction((tx) =>
      swapEgressScheduled({
        block: block as any,
        event: event as any,
        prisma: tx,
      }),
    );

    const swap = await prisma.swap.findFirstOrThrow({
      where: { swapDepositChannelId: swapDepositChannel.id },
      include: {
        egress: {
          select: {
            amount: true,
            scheduledAt: true,
            chain: true,
          },
        },
        fees: true,
      },
    });

    expect(swap).toMatchSnapshot({
      id: expect.any(BigInt),
      egressId: expect.any(BigInt),
      createdAt: expect.any(Date),
      updatedAt: expect.any(Date),
      swapDepositChannelId: expect.any(BigInt),
      fees: [{ id: expect.any(BigInt), swapId: expect.any(BigInt) }],
    });
    expect(axios.post).not.toHaveBeenCalledWith(
      expect.anything(),
      expect.objectContaining({ method: 'cf_swap_rate' }),
    );
  });

  it('updates an existing swap when the egress asset is not the native asset', async () => {
    const { swapId } = event.args;

    await prisma.egress.create({
      data: {
        chain: event.args.egressId[0].__kind,
        nativeId: BigInt(event.args.egressId[1]),
        amount: '1234567890',
        scheduledAt: new Date(block.timestamp),
        scheduledBlockIndex: `${block.height}-${event.indexInBlock}`,
      },
    });

    // store a new swap intent to initiate a new swap
    const swapDepositChannel = await createDepositChannel({
      swaps: {
        create: {
          nativeId: BigInt(swapId),
          depositAmount: '10000000000',
          swapInputAmount: '10000000000',
          swapOutputAmount: '10000000000',
          depositReceivedAt: new Date(block.timestamp - 12000),
          depositReceivedBlockIndex: `${block.height - 100}-${
            event.indexInBlock
          }`,
          swapExecutedAt: new Date(block.timestamp - 6000),
          swapExecutedBlockIndex: `${block.height}-${event.indexInBlock}`,
          srcAsset: Assets.ETH,
          destAsset: Assets.USDC,
          destAddress: ETH_ADDRESS,
          type: 'SWAP',
        },
      },
    });

    await prisma.$transaction((tx) =>
      swapEgressScheduled({
        block: block as any,
        event: event as any,
        prisma: tx,
      }),
    );

    const swap = await prisma.swap.findFirstOrThrow({
      where: { swapDepositChannelId: swapDepositChannel.id },
      include: {
        egress: {
          select: {
            amount: true,
            scheduledAt: true,
            chain: true,
          },
        },
        fees: true,
      },
    });

    expect(swap).toMatchSnapshot({
      id: expect.any(BigInt),
      egressId: expect.any(BigInt),
      createdAt: expect.any(Date),
      updatedAt: expect.any(Date),
      swapDepositChannelId: expect.any(BigInt),
      fees: [{ id: expect.any(BigInt), swapId: expect.any(BigInt) }],
    });
    expect(
      jest.mocked(axios.post).mock.calls.map((call) => call[1]),
    ).toMatchSnapshot();
  });

  it('uses the fee from the event and creates an egress record', async () => {
    const { swapId } = event.args;

    // store a new swap intent to initiate a new swap
    const swapDepositChannel = await createDepositChannel({
      swaps: {
        create: {
          nativeId: BigInt(swapId),
          depositAmount: '10000000000',
          swapInputAmount: '10000000000',
          swapOutputAmount: '10000000000',
          depositReceivedAt: new Date(block.timestamp - 12000),
          depositReceivedBlockIndex: `${block.height - 100}-${
            event.indexInBlock
          }`,
          swapExecutedAt: new Date(block.timestamp - 6000),
          swapExecutedBlockIndex: `${block.height}-${event.indexInBlock}`,
          srcAsset: Assets.ETH,
          destAsset: Assets.DOT,
          destAddress: DOT_ADDRESS,
          type: 'SWAP',
        },
      },
    });

    await prisma.$transaction((tx) =>
      swapEgressScheduled({
        block: block as any,
        event: {
          ...event,
          args: {
            ...event.args,
            fee: '0x123',
            amount: '0x1234567890',
          },
        },
        prisma: tx,
      }),
    );

    const swap = await prisma.swap.findFirstOrThrow({
      where: { swapDepositChannelId: swapDepositChannel.id },
      include: {
        egress: {
          select: {
            amount: true,
            scheduledAt: true,
            chain: true,
          },
        },
        fees: true,
      },
    });

    expect(swap).toMatchSnapshot({
      id: expect.any(BigInt),
      egressId: expect.any(BigInt),
      createdAt: expect.any(Date),
      updatedAt: expect.any(Date),
      swapDepositChannelId: expect.any(BigInt),
      fees: [{ id: expect.any(BigInt), swapId: expect.any(BigInt) }],
    });
    expect(axios.post).not.toHaveBeenCalledWith(
      expect.anything(),
      expect.objectContaining({ method: 'cf_swap_rate' }),
    );
  });
});
