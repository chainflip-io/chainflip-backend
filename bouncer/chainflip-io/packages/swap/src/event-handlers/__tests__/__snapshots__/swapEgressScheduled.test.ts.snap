// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`swapEgressScheduled updates an existing swap when the egress amount is higher than the swapOutputAmount 1`] = `
{
  "ccmDepositReceivedBlockIndex": null,
  "ccmGasBudget": null,
  "ccmMessage": null,
  "createdAt": Any<Date>,
  "depositAmount": "1000",
  "depositReceivedAt": 2022-12-06T14:31:33.000Z,
  "depositReceivedBlockIndex": "20-0",
  "destAddress": "1yMmfLti1k3huRQM2c47WugwonQMqTvQ2GUFxnU7Pcs7xPo",
  "destAsset": "DOT",
  "egress": {
    "amount": "0",
    "chain": "Ethereum",
    "scheduledAt": 2022-12-06T14:31:45.000Z,
  },
  "egressId": Any<BigInt>,
  "fees": [
    {
      "amount": "1000",
      "asset": "DOT",
      "id": Any<BigInt>,
      "swapId": Any<BigInt>,
      "type": "EGRESS",
    },
  ],
  "id": Any<BigInt>,
  "intermediateAmount": null,
  "nativeId": 9876545n,
  "srcAsset": "ETH",
  "swapDepositChannelId": Any<BigInt>,
  "swapExecutedAt": 2022-12-06T14:31:39.000Z,
  "swapExecutedBlockIndex": "120-0",
  "swapInputAmount": "1000",
  "swapOutputAmount": "1000",
  "txHash": null,
  "type": "SWAP",
  "updatedAt": Any<Date>,
}
`;

exports[`swapEgressScheduled updates an existing swap when the egress asset is not the native asset 1`] = `
{
  "ccmDepositReceivedBlockIndex": null,
  "ccmGasBudget": null,
  "ccmMessage": null,
  "createdAt": Any<Date>,
  "depositAmount": "10000000000",
  "depositReceivedAt": 2022-12-06T14:31:33.000Z,
  "depositReceivedBlockIndex": "20-0",
  "destAddress": "0x6Aa69332B63bB5b1d7Ca5355387EDd5624e181F2",
  "destAsset": "USDC",
  "egress": {
    "amount": "9976511899",
    "chain": "Ethereum",
    "scheduledAt": 2022-12-06T14:31:45.000Z,
  },
  "egressId": Any<BigInt>,
  "fees": [
    {
      "amount": "23488101",
      "asset": "USDC",
      "id": Any<BigInt>,
      "swapId": Any<BigInt>,
      "type": "EGRESS",
    },
  ],
  "id": Any<BigInt>,
  "intermediateAmount": null,
  "nativeId": 9876545n,
  "srcAsset": "ETH",
  "swapDepositChannelId": Any<BigInt>,
  "swapExecutedAt": 2022-12-06T14:31:39.000Z,
  "swapExecutedBlockIndex": "120-0",
  "swapInputAmount": "10000000000",
  "swapOutputAmount": "10000000000",
  "txHash": null,
  "type": "SWAP",
  "updatedAt": Any<Date>,
}
`;

exports[`swapEgressScheduled updates an existing swap when the egress asset is not the native asset 2`] = `
[
  {
    "id": 1,
    "jsonrpc": "2.0",
    "method": "cf_swap_rate",
    "params": [
      "ETH",
      "USDC",
      "0x777777",
      undefined,
    ],
  },
]
`;

exports[`swapEgressScheduled updates an existing swap when the egress asset is the native asset 1`] = `
{
  "ccmDepositReceivedBlockIndex": null,
  "ccmGasBudget": null,
  "ccmMessage": null,
  "createdAt": Any<Date>,
  "depositAmount": "10000000000",
  "depositReceivedAt": 2022-12-06T14:31:33.000Z,
  "depositReceivedBlockIndex": "20-0",
  "destAddress": "1yMmfLti1k3huRQM2c47WugwonQMqTvQ2GUFxnU7Pcs7xPo",
  "destAsset": "DOT",
  "egress": {
    "amount": "9992170633",
    "chain": "Ethereum",
    "scheduledAt": 2022-12-06T14:31:45.000Z,
  },
  "egressId": Any<BigInt>,
  "fees": [
    {
      "amount": "7829367",
      "asset": "DOT",
      "id": Any<BigInt>,
      "swapId": Any<BigInt>,
      "type": "EGRESS",
    },
  ],
  "id": Any<BigInt>,
  "intermediateAmount": null,
  "nativeId": 9876545n,
  "srcAsset": "ETH",
  "swapDepositChannelId": Any<BigInt>,
  "swapExecutedAt": 2022-12-06T14:31:39.000Z,
  "swapExecutedBlockIndex": "120-0",
  "swapInputAmount": "10000000000",
  "swapOutputAmount": "10000000000",
  "txHash": null,
  "type": "SWAP",
  "updatedAt": Any<Date>,
}
`;

exports[`swapEgressScheduled uses the fee from the event and creates an egress record 1`] = `
{
  "ccmDepositReceivedBlockIndex": null,
  "ccmGasBudget": null,
  "ccmMessage": null,
  "createdAt": Any<Date>,
  "depositAmount": "10000000000",
  "depositReceivedAt": 2022-12-06T14:31:33.000Z,
  "depositReceivedBlockIndex": "20-0",
  "destAddress": "1yMmfLti1k3huRQM2c47WugwonQMqTvQ2GUFxnU7Pcs7xPo",
  "destAsset": "DOT",
  "egress": {
    "amount": "78187493520",
    "chain": "Ethereum",
    "scheduledAt": 2022-12-06T14:31:45.000Z,
  },
  "egressId": Any<BigInt>,
  "fees": [
    {
      "amount": "291",
      "asset": "DOT",
      "id": Any<BigInt>,
      "swapId": Any<BigInt>,
      "type": "EGRESS",
    },
  ],
  "id": Any<BigInt>,
  "intermediateAmount": null,
  "nativeId": 9876545n,
  "srcAsset": "ETH",
  "swapDepositChannelId": Any<BigInt>,
  "swapExecutedAt": 2022-12-06T14:31:39.000Z,
  "swapExecutedBlockIndex": "120-0",
  "swapInputAmount": "10000000000",
  "swapOutputAmount": "10000000000",
  "txHash": null,
  "type": "SWAP",
  "updatedAt": Any<Date>,
}
`;
